### PR TITLE
Replace list-spiders with shub-image-info

### PR DESCRIPTION
**Pull request by vshlapakov**
We're introducing a new command to Kumo contract: it's called `shub-image-info` and will deprecate current `list-spiders`. Comparing with `list-spiders` command - the new command should return JSON encoded metadata for a given image, incl. a project type, spiders metadata and probably some other fields in the future. As now we're changing the way how arguments and settings are being passed to a spider/script, scripts become a Scrapy-specific feature, and should be included into the same spiders list with `py:` prefix.

Important note, that we must provide backward compatibility, so if `shub-image-info` is not found, there's fall back to `list-spiders` with a deprecation warning.

Please, review.

### CODE REVIEW

**Comment by chekunkov** [comment]
Tested with https://github.com/scrapinghub/custom-images-examples/pull/9, LGTM in general

File: shub/image/list.py [hunk 1]
```diff
@@ -52,27 +54,38 @@ def list_cmd_full(target, silent, version):
...
-    if status_code != 0:
+    environment = {'JOB_SETTINGS': json.dumps(settings)}
+    exit_code, logs = _run_cmd_in_docker_container(
+            image_name, 'shub-image-info', environment)
+    if exit_code == 0:
+        return _extract_metadata_from_image_info_output(logs)
+    # shub-image-info command not found, fallback to list-spiders
+    elif exit_code == 127:
+        # FIXME we should pass some value for SCRAPY_PROJECT_ID anyway
+        # to handle `scrapy list` cmd properly via sh_scrapy entrypoint
```
Comment (jdemaeyer) [inline shub/image/list.py]: (I know this `FIXME` was just copied over, but) What happens when there is no/an empty `SCRAPY_PROJECT_ID` defined, as implemented?

File: shub/image/list.py [hunk 2]
```diff
@@ -89,27 +102,47 @@ def _get_project_settings(project, endpoint, apikey):
...
+        raise_shub_image_info_error('output is not a valid JSON dict')
+    if not isinstance(spiders_list, list):
+        raise_shub_image_info_error('spiders section must be a list')
+
+    spiders, scripts = [], []
+    for name in spiders_list:
+        if not (name and isinstance(name, string_types)):
+            raise_shub_image_info_error("spider name can't be empty or non-string")
+        scripts.append(name[3:]) if name.startswith('py:') else spiders.append(name)
+    return {'spiders': spiders, 'scripts': scripts}
```
Comment (jdemaeyer) [inline shub/image/list.py]: For `list-spiders` (and in the fallback to it in line 84 above), we ran the spider list through `utils.valid_spiders()`. Is there a reason to skip it here?

Comment (vshlapakov) [inline shub/image/list.py]: No reason, it's a good idea to validate spiders/scripts lists with the same regex, thanks :+1:

File: shub/image/list.py (hunk 1 above)
Comment (vshlapakov) [inline shub/image/list.py]: Earlier `scrapinghub-entrypoint-scrapy` relied on `SCRAPY_PROJECT_ID` to decide between a spider and a python script, later we changed it with https://github.com/scrapinghub/scrapinghub-entrypoint-scrapy/pull/26 in version `0.8.8`, but there may be a couple of custom images in-use relying on previous sh-ep-scrapy versions. So if there's no `SCAPY_PROJECT_ID` for such cases, `scrapy list` will be launched incorrectly (as a package script) and fail with not-found-error or something like this.

Now I wonder how many projects use custom images with oldy `scrapinghub-entrypoint-scrapy` versions in real, maybe it's already fine to finally drop it. That's a shame that PyPi disabled download stats :confused:

File: shub/image/list.py (hunk 2 above)
Comment (vshlapakov) [inline shub/image/list.py]: Addressed with https://github.com/scrapinghub/shub/pull/279/commits/e8e0989f66fad911287728c13b9ba5458a7e66cc

File: shub/image/list.py (hunk 1 above)
Comment (jdemaeyer) [inline shub/image/list.py]: Hm I see. Let's wait for #269 to drop it

**Comment by vshlapakov** [comment]
Thanks @jdemaeyer! :slightly_smiling_face: Additional note - I believe we should release `shub` with the changes before we could merge https://github.com/scrapinghub/custom-images-examples/issues/9, are there any ETA for next `2.6.2`? It's not urgent, just to get some estimation.

**Pull request merged**